service/storage_service.cc

```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+                                slogger.debug("decommission[{}]: Got heartbeat response from node={}", uuid, node);
+                                return make_ready_future<>();
+                            });
+                        }).handle_exception([uuid] (std::exception_ptr ep) {
+                            slogger.warn("decommission[{}]: Failed to send heartbeat", uuid);
+                        }).get();
+                        int nr_seconds = 10;
+                        while (!(*heartbeat_updater_done) && nr_seconds--) {
+                            sleep_abortable(std::chrono::seconds(1), ss._abort_source).get();
+                        }
+                    }
+                    slogger.debug("decommission[{}]: Stopped heartbeat_updater", uuid);
+                });
+                auto stop_heartbeat_updater = defer([&] {
+                    *heartbeat_updater_done = true;
+                    heartbeat_updater.get();
+                });
 
-            slogger.info("DECOMMISSIONING: unbootstrap starts");
-            ss.unbootstrap();
-            slogger.info("DECOMMISSIONING: unbootstrap done");
+                // Step 5: Start to sync data
+                slogger.info("DECOMMISSIONING: unbootstrap starts");
+                ss.unbootstrap();
+                slogger.info("DECOMMISSIONING: unbootstrap done");
+
+                // Step 6: Finish
+                req.cmd = node_ops_cmd::decommission_done;
+                parallel_for_each(nodes, [&ss, &req, uuid] (const gms::inet_address& node) {
+                    return ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
+                        slogger.debug("decommission[{}]: Got done response from node={}", uuid, node);
+                        return make_ready_future<>();
+                    });
+                }).get();
+                slogger.info("decommission[{}]: Finished decommission operation, removing node={}, sync_nodes={}, ignore_nodes={}", uuid, endpoint, nodes, ignore_nodes);
+            } catch (...) {
+                slogger.warn("decommission[{}]: Abort decommission operation started, removing node={}, sync_nodes={}, ignore_nodes={}", uuid, endpoint, nodes, ignore_nodes);
+                // we need to revert the effect of prepare verb the decommission ops is failed
+                req.cmd = node_ops_cmd::decommission_abort;
+                parallel_for_each(nodes, [&ss, &req, &nodes_unknown_verb, &nodes_down, uuid] (const gms::inet_address& node) {
+                    if (nodes_unknown_verb.contains(node) || nodes_down.contains(node)) {
+                        // No need to revert previous prepare cmd for those who do not apply prepare cmd.
+                        return make_ready_future<>();
+                    }
+                    return ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
```

Comment:
Is it OK to send an abort command to nodes you already succeeded sending decommission_done?
(I hope it is, given you send a uuid with it too).

If not, you can keep of list of those and ignore them too along with nodes_unknown_verb/down.

Reply:
It is ok.